pkg/registry/core/pod/strategy_test.go

```diff
@@ -1435,3 +1440,89 @@ func TestDropNonEphemeralContainerUpdates(t *testing.T) {
 		})
 	}
 }
+
+func TestNodeInclusionPolicyEnablement(t *testing.T) {
+	var (
+		honor            = api.NodeInclusionPolicyHonor
+		ignore           = api.NodeInclusionPolicyIgnore
+		emptyConstraints = []api.TopologySpreadConstraint{
+			{
+				WhenUnsatisfiable: api.DoNotSchedule,
+				TopologyKey:       "kubernetes.io/hostname",
+				MaxSkew:           1,
+			},
+		}
+		defaultConstraints = []api.TopologySpreadConstraint{
+			{
+				NodeAffinityPolicy: &honor,
+				NodeTaintsPolicy:   &ignore,
+				WhenUnsatisfiable:  api.DoNotSchedule,
+				TopologyKey:        "kubernetes.io/hostname",
+				MaxSkew:            1,
+			},
+		}
+	)
+
+	tests := []struct {
+		name                          string
+		topologySpreadConstraints     []api.TopologySpreadConstraint
+		wantTopologySpreadConstraints []api.TopologySpreadConstraint
+		enableNodeInclusionPolicy     bool
+	}{
+		{
+			name:                          "nodeInclusionPolicy enabled with topology unset",
+			topologySpreadConstraints:     emptyConstraints,
+			wantTopologySpreadConstraints: emptyConstraints,
+			enableNodeInclusionPolicy:     true,
+		},
+		{
+			name:                          "nodeInclusionPolicy enabled with topology configured",
+			topologySpreadConstraints:     defaultConstraints,
+			wantTopologySpreadConstraints: defaultConstraints,
+			enableNodeInclusionPolicy:     true,
+		},
+		{
+			name:                          "nodeInclusionPolicy disabled with topology configured",
+			topologySpreadConstraints:     defaultConstraints,
+			wantTopologySpreadConstraints: emptyConstraints,
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.NodeInclusionPolicyInPodTopologySpread, tc.enableNodeInclusionPolicy)()
+
+			pod := &api.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Namespace: "default",
+					Name:      "foo",
+				},
+				Spec: api.PodSpec{
+					RestartPolicy: api.RestartPolicyAlways,
+					DNSPolicy:     api.DNSDefault,
+					Containers: []api.Container{
+						{
+							Name:                     "container",
+							Image:                    "image",
+							ImagePullPolicy:          "IfNotPresent",
+							TerminationMessagePolicy: "File",
+						},
+					},
+				},
+			}
+			wantPod := pod.DeepCopy()
+			pod.Spec.TopologySpreadConstraints = append(pod.Spec.TopologySpreadConstraints, tc.topologySpreadConstraints...)
+
+			errs := Strategy.Validate(genericapirequest.NewContext(), pod)
+			if len(errs) != 0 {
+				t.Errorf("Unexpected error: %v", errs.ToAggregate())
+			}
+
+			Strategy.PrepareForCreate(genericapirequest.NewContext(), pod)
```

Comment:
We'll call `DropDisabledPodFields` in `PrepareForCreate` which will processes below logic and helps us to verify the feature gate enablement. https://github.com/kubernetes/kubernetes/blob/ecdbb0258049b9c902b53c77d386b741e2a25cb7/pkg/api/pod/util.go#L596-L609

Reply:
This is good test, but it's still not exactly what I wanted.
The crucial part of https://github.com/kubernetes/kubernetes/pull/97058/files#diff-7826f7adbc1996a05ab52e3f5f02429e94b68ce6bce0dc534d1be636154fded3R246-R282 that I linked you is the fact that the feature gate is changed in the middle of the test, so basically:
- we store an object with a new field
- we disabled the FG
- we check if updating the object with the FG now disabled works as expected now..

Reply:
Kindly ping @wojtek-t , added a new test as advised.